### PR TITLE
[23-01-21] 문자열 나누기 (1단계)

### PR DESCRIPTION
**Pull request by Jeong-jeong**
## 문제
<p style="box-sizing: border-box; padding: 0px; margin: 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; overflow-wrap: break-word; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">문자열<span> </span><code style="box-sizing: border-box; padding: 2px 6px; margin: 0px 2px 0px 0px; color: rgb(205, 215, 224); font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 14px; line-height: inherit; font-family: Hack, Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif; overflow-wrap: break-word; border: 1px solid rgb(23, 35, 52); background-color: rgb(32, 43, 61); border-radius: 0.25rem;">s</code>가 입력되었을 때 다음 규칙을 따라서 이 문자열을 여러 문자열로 분해하려고 합니다.</p><ul style="box-sizing: border-box; padding: 0px 0px 0px 2.5rem; margin: 1rem 0px 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: inherit; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 16px; line-height: 1.6; font-family: inherit; letter-spacing: -0.009em; list-style: inherit;">먼저 첫 글자를 읽습니다. 이 글자를 x라고 합시다.</li><li style="box-sizing: border-box; padding: 0px; margin: 0.5rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 16px; line-height: 1.6; font-family: inherit; letter-spacing: -0.009em; list-style: inherit;">이제 이 문자열을 왼쪽에서 오른쪽으로 읽어나가면서, <strong>x와 x가 아닌 다른 글자들이 나온 횟수를 각각 셉니다.</strong> 처음으로 두 횟수가 같아지는 순간 멈추고, 지금까지 읽은 문자열을 분리합니다.</li><li style="box-sizing: border-box; padding: 0px; margin: 0.5rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 16px; line-height: 1.6; font-family: inherit; letter-spacing: -0.009em; list-style: inherit;"><code style="box-sizing: border-box; padding: 2px 6px; margin: 0px 2px 0px 0px; color: rgb(205, 215, 224); font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 14px; line-height: inherit; font-family: Hack, Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif; overflow-wrap: break-word; border: 1px solid rgb(23, 35, 52); background-color: rgb(32, 43, 61); border-radius: 0.25rem;">s</code>에서 분리한 문자열을 빼고 남은 부분에 대해서 이 과정을 반복합니다. 남은 부분이 없다면 종료합니다.</li><li style="box-sizing: border-box; padding: 0px; margin: 0.5rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 16px; line-height: 1.6; font-family: inherit; letter-spacing: -0.009em; list-style: inherit;">만약 두 횟수가 다른 상태에서 더 이상 읽을 글자가 없다면, 역시 지금까지 읽은 문자열을 분리하고, 종료합니다.</li></ul><p style="box-sizing: border-box; padding: 0px; margin: 0.75rem 0px 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; overflow-wrap: break-word; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">문자열<span> </span><code style="box-sizing: border-box; padding: 2px 6px; margin: 0px 2px 0px 0px; color: rgb(205, 215, 224); font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 14px; line-height: inherit; font-family: Hack, Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif; overflow-wrap: break-word; border: 1px solid rgb(23, 35, 52); background-color: rgb(32, 43, 61); border-radius: 0.25rem;">s</code>가 매개변수로 주어질 때, 위 과정과 같이 문자열들로 분해하고, 분해한 문자열의 개수를 return 하는 함수 solution을 완성하세요.</p><hr style="box-sizing: content-box; padding: 0px; margin: 1.5rem 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: inherit; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; height: 0px; overflow: visible; border-width: 0.0625rem 0px 0px; border-right-style: initial; border-bottom-style: initial; border-left-style: initial; border-color: rgb(23, 35, 52); border-image: initial; border-top-style: solid; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h5 style="box-sizing: border-box; padding: 0px; margin: 1.5rem 0px 0px; color: white; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 15px; line-height: 1.5; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.009em; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">제한사항</h5><ul style="box-sizing: border-box; padding: 0px 0px 0px 2.5rem; margin: 0.75rem 0px 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: inherit; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 16px; line-height: 1.6; font-family: inherit; letter-spacing: -0.009em; list-style: inherit;">1 ≤<span> </span><code style="box-sizing: border-box; padding: 2px 6px; margin: 0px 2px 0px 0px; color: rgb(205, 215, 224); font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 14px; line-height: inherit; font-family: Hack, Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif; overflow-wrap: break-word; border: 1px solid rgb(23, 35, 52); background-color: rgb(32, 43, 61); border-radius: 0.25rem;">s</code>의 길이 ≤ 10,000</li><li style="box-sizing: border-box; padding: 0px; margin: 0.5rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 16px; line-height: 1.6; font-family: inherit; letter-spacing: -0.009em; list-style: inherit;"><code style="box-sizing: border-box; padding: 2px 6px; margin: 0px 2px 0px 0px; color: rgb(205, 215, 224); font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 14px; line-height: inherit; font-family: Hack, Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif; overflow-wrap: break-word; border: 1px solid rgb(23, 35, 52); background-color: rgb(32, 43, 61); border-radius: 0.25rem;">s</code>는 영어 소문자로만 이루어져 있습니다.</li></ul><hr style="box-sizing: content-box; padding: 0px; margin: 1.5rem 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: inherit; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; height: 0px; overflow: visible; border-width: 0.0625rem 0px 0px; border-right-style: initial; border-bottom-style: initial; border-left-style: initial; border-color: rgb(23, 35, 52); border-image: initial; border-top-style: solid; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h5 style="box-sizing: border-box; padding: 0px; margin: 1.5rem 0px 0px; color: white; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 15px; line-height: 1.5; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.009em; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">입출력 예</h5>

s | result
-- | --
"banana" | 3
"abracadabra" | 6
"aaabbaccccabba" | 3

<hr style="box-sizing: content-box; padding: 0px; margin: 1.5rem 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: inherit; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; height: 0px; overflow: visible; border-width: 0.0625rem 0px 0px; border-right-style: initial; border-bottom-style: initial; border-left-style: initial; border-color: rgb(23, 35, 52); border-image: initial; border-top-style: solid; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><h5 style="box-sizing: border-box; padding: 0px; margin: 1.5rem 0px 0px; color: white; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 15px; line-height: 1.5; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.009em; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">입출력 예 설명</h5><p style="box-sizing: border-box; padding: 0px; margin: 0.75rem 0px 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; overflow-wrap: break-word; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">입출력 예 #1<br style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font: inherit;"><code style="box-sizing: border-box; padding: 2px 6px; margin: 0px 2px 0px 0px; color: rgb(205, 215, 224); font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 14px; line-height: inherit; font-family: Hack, Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif; overflow-wrap: break-word; border: 1px solid rgb(23, 35, 52); background-color: rgb(32, 43, 61); border-radius: 0.25rem;">s</code>="banana"인 경우 ba - na - na와 같이 분해됩니다.</p><p style="box-sizing: border-box; padding: 0px; margin: 0.75rem 0px 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; overflow-wrap: break-word; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">입출력 예 #2<br style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font: inherit;"><code style="box-sizing: border-box; padding: 2px 6px; margin: 0px 2px 0px 0px; color: rgb(205, 215, 224); font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 14px; line-height: inherit; font-family: Hack, Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif; overflow-wrap: break-word; border: 1px solid rgb(23, 35, 52); background-color: rgb(32, 43, 61); border-radius: 0.25rem;">s</code>="abracadabra"인 경우 ab - ra - ca - da - br - a와 같이 분해됩니다.</p><p style="box-sizing: border-box; padding: 0px; margin: 0.75rem 0px 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; overflow-wrap: break-word; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">입출력 예 #3<br style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font: inherit;"><code style="box-sizing: border-box; padding: 2px 6px; margin: 0px 2px 0px 0px; color: rgb(205, 215, 224); font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 14px; line-height: inherit; font-family: Hack, Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif; overflow-wrap: break-word; border: 1px solid rgb(23, 35, 52); background-color: rgb(32, 43, 61); border-radius: 0.25rem;">s</code>="aaabbaccccabba"인 경우 aaabbacc - ccab - ba와 같이 분해됩니다.</p>

## 푸는데 걸린 시간

40분 

## 깨달은 점 (필요한 경우)

x와 x가 아닌 다른 글자들이 나온 횟수를 각각 셉니다.  -> x가 아닌 글자들도 각각 세는게 아니라, x 세고, x가 아닌 글자들을 합쳐서 센다

## 스크린샷 (필요할 경우)

image

## 관련 문서 링크 (필요할 경우)

https://school.programmers.co.kr/learn/courses/30/lessons/140108

### CODE REVIEW

File: jeongs/jan/23-01-21-문자열나누기.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    let target = array[0];
+    let counts = [0, 0];
+    let result = 0;
+    
+    array.forEach((str, idx) => {
+        if (counts[0] === counts[1] && counts[0] > 0 && counts[1] > 0) {
+            result+= 1;
+            target = str;
+            counts = [0, 0];
+        }
```
Comment (Jeong-jeong) [inline jeongs/jan/23-01-21-문자열나누기.js:12]: 순회할 때마다, xCount와 x가 아닌 숫자 count를 비교.
둘이 같으면 resut + 1, x를 현재값으로, counts도 초기화.

File: jeongs/jan/23-01-21-문자열나누기.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+        (str === target) ? counts[0] += 1 : counts[1] += 1;
```
Comment (Jeong-jeong) [inline jeongs/jan/23-01-21-문자열나누기.js:14]: 현재값과 x가 같다면? xCount를 증가, 현재값과 x가 다를 경우 x가 아닌 숫자들 카운트 증가.

File: jeongs/jan/23-01-21-문자열나누기.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    })
+    return result + 1;
```
Comment (Jeong-jeong) [inline jeongs/jan/23-01-21-문자열나누기.js:16]: 마지막 놈은 카운트가 안되므로, + 1

**Review by sobaman — APPROVED** [review]
이것은 나의 머리로 이해하기에는 아직 역부족입니다. 숙련자가 가질 수 있는 것

**Comment by sobaman** [comment]
번외로, 병합 풀 요청 버튼은 마치 나를 유혹하는 것 처럼 생겼습니다

**Comment by Jeong-jeong** [comment]
@sobaman 또 누르면 죽일 것입니다

**Pull request merged**